backend/src/v4/models/mitigation.js

```diff
@@ -132,11 +176,103 @@ class Mitigation {
 			mitigations[i] = newMitigation;
 		}
 
-		await this.clearAll(account);
+		if (clearAll) {
+			await this.clearAll(account);
+		}
+
 		await mitigationColl.insertMany(mitigations);
 
 		return mitigations;
 	}
+
+	async update(account, id, updateData) {
+		await db.updateOne(account, "mitigations", { _id: id }, updateData);
+	}
+
+	async addRiskRefToMitigation(account, mitigation, riskReference) {
+		const referencedRisks = mitigation.referencedRisks;
+
+		if (!referencedRisks.includes(riskReference)) {
+			referencedRisks.push(riskReference);
+			await this.update(account, mitigation._id, { $set: { referencedRisks } });
+		}
+	}
+
+	async updateMitigationsFromRisk(account, model, oldRisk, updatedRisk) {
+		const riskId = updatedRisk._id;
+		const oldStatusIsResolved = !!oldRisk?.mitigation_desc && isMitigationStatusResolved(oldRisk.mitigation_status);
+		const newStatusIsResolved = !!updatedRisk?.mitigation_desc && isMitigationStatusResolved(updatedRisk.mitigation_status);
+
+		// if risk was and remains unresolved
+		if (!oldStatusIsResolved && !newStatusIsResolved) {
+			return;
+		}
+
+		const mitigationDetails = {
+			mitigation_desc: updatedRisk.mitigation_desc,
+			mitigation_detail: updatedRisk.mitigation_detail,
+			mitigation_stage: updatedRisk.mitigation_stage,
+			mitigation_type: updatedRisk.mitigation_type,
+			mitigation_status: updatedRisk.mitigation_status,
+			category: updatedRisk.category,
+			location_desc: updatedRisk.location_desc,
+			element: updatedRisk.element,
+			risk_factor: updatedRisk.risk_factor,
+			scope: updatedRisk.scope,
+			associated_activity: updatedRisk.associated_activity
+		};
+		const formattedReference = formatRiskReference(account, model, riskId);
+
+		// if risk becomes resolved
+		if (!oldStatusIsResolved && newStatusIsResolved) {
+			const mitigations = await this.findMitigationSuggestions(account, mitigationDetails, []);
```

Comment:
This will be prone to errors as `findMitigationSuggestions` has a more relaxed criteria search than we we want here

for e.g. if this risk has {element: "abc"} and the rest are undefined, it will return things like `{ element: "abc", category: "xyz"}` etc. which is not an exact match to the rule.

we want to do a find here where `{element: "abc", location_desc: null, risk_factor: null }` so we only get the exact match.

Then you can also do a projection for referencedRisk so you don't get the whole document.